### DISCUSSION TITLE
Resource Proxies

### DISCUSSION THREAD

**Discussion by mrpedrobraga** [godotengine/godot-proposals#6314] [General Discussions]
There's something that I'd like to turn into a proposal but would like to discuss first,

In my framework I needed a way to refer to a resource on disk without loading it. I did so by implementing `ResourceProxy : Resource` which stores the file and can be asked `get_resource()` to retrieve the resource, `load`, `preload`, too.

It cuts the work a lot by making it basically a `RefCounted` string and I only have to update it once. Still, given that Godot has the ability of updating resource paths for exported resources on scenes when you move resources in the file system, it would be great to have the editor watch over a file as I move it around, but the scene to not load it when instantiated.

**Comment by Daylily-Zeleen** [comment]
It seem just a object which container a resource path as peoperty, and provide a series of methods to access the target resource, right?

I think it is easy to be implemented, but it is necessary?

If this class it used in editor, in order to save it's instacne in a scne file(*.tscn/ *.scn), it should be a resource, too.
if not, why not use a String directly?

**Reply by mrpedrobraga** [reply]
Strings don't get handled when files are moved or renamed around on the FileSystem. The Editor does watch out for `Resource`s though.

**Reply by Daylily-Zeleen** [reply]
ahh, got it.

Implements a resource class `ResourceProxy` or `ResourceReference` is the most easy way.

Here is another plan，I think we should add a new `PropertyHint`, such as `PROPERTY_HINT_RESOURCE_PATH` to hint a String property that it is a resource path, and implement the logic to modify this property automatically when its target resource path change.

**Reply by hawkerm** [reply]
I too was just looking for something akin to `NodePath` for this as well, but that would let me easily select any `res://` node within my project without me having to copy the string from it and paste it in as an exported `string`.

The updating if I move/rename the resource would be great value to have it tracked by the editor.

I then pass this around in my code as a 'reference' to the resource and have a different service in my code cache/load/manage these assets, be them sounds or effect scenes to spawn, etc...

**Comment by mrpedrobraga** [comment]
@Daylily-Zeleen A PropertyHint would work just as well; But I'm unaware of how the editor would keep track of this.

During the implementation of `PROPERTY_HINT_NODE_TYPE` there were concerns regarding the implementation, and ultimately it was implemented by editing a hidden metadata to store the node path in the scene;

Perhaps something similar can be done for a `PROPERTY_HINT_RESOURCE_TYPE_PATH` and store the [Resource UID](https://docs.godotengine.org/en/stable/classes/class_resourceuid.html) in it?

Then, if `@GDScript.load` can load resources using their UID, we can get the behaviour of referencing without loading.

**Reply by Daylily-Zeleen** [reply]
Please try it, and open a proposal.
